# Examen 1.
En este primer examen se evaluan los conceptos básicos de Python a través de la resolución de ejercicios que van de menos a más. Se revisan temas de:
* Definición de variables.
* Tipos de variables
* Arreglos de numpy simples
* Funciones.

 <p xmlns:cc="http://creativecommons.org/ns#" xmlns:dct="http://purl.org/dc/terms/">Examen 1 by <a rel="cc:attributionURL dct:creator" property="cc:attributionName" href="https://www.macti.unam.mx">Luis M. de la Cruz</a> is licensed under <a href="http://creativecommons.org/licenses/by-sa/4.0/?ref=chooser-v1" target="_blank" rel="license noopener noreferrer" style="display:inline-block;">Attribution-ShareAlike 4.0 International<img style="height:22px!important;margin-left:3px;vertical-align:text-bottom;" src="https://mirrors.creativecommons.org/presskit/icons/cc.svg?ref=chooser-v1"><img style="height:22px!important;margin-left:3px;vertical-align:text-bottom;" src="https://mirrors.creativecommons.org/presskit/icons/by.svg?ref=chooser-v1"><img style="height:22px!important;margin-left:3px;vertical-align:text-bottom;" src="https://mirrors.creativecommons.org/presskit/icons/sa.svg?ref=chooser-v1"></a></p> 

In [1]:
from macti.evaluation import *
quizz = Quizz('q1', 'HeCompA')

<div class="alert alert-success">

## **Ejercicio 1.**

<font color="DarkBlue">
    
En la siguiente celda de código, define las siguientes variables:

```python
k = 1000
L = 0.5
nombre = 'luis mi'
```

Luego contesta las siguientes tres preguntas.

**Nota**: escribe tu respuesta entre comillas, por ejemplo: `respuesta = "b"`

</font>
</div>

In [2]:
# Define las varibles
# k = ...
### BEGIN SOLUTION

### FileAnswer definition
file_answer = FileAnswer() 

k = 1000
L = 0.5
nombre = 'luis mi'
### END SOLUTION

**1**. ¿De qué tipo es la variable `k`?

    a. <class 'str'>
    b. <class 'complex'>
    c. <class 'float'>
    d. <class 'list'>
    e. <class 'int'>

In [3]:
# respuesta = "..."
### BEGIN SOLUTION
respuesta = "e"
file_answer.write('1', respuesta, 'Utiliza la función type(k) para revisar el tipo de la variable.')
### END SOLUTION

In [4]:
quizz.eval_option('1', respuesta)

----------------------------------------
Tu respuesta: e, es correcta.
----------------------------------------


**2**. ¿De qué tipo es la varible `L`?

    a. <class 'str'>
    b. <class 'complex'>
    c. <class 'float'>
    d. <class 'list'>
    e. <class 'int'>

In [5]:
# respuesta = "..."
### BEGIN SOLUTION
respuesta = "c"
file_answer.write('2', respuesta, 'Utiliza la función type(L) para revisar el tipo de la variable.')
### END SOLUTION

In [6]:
quizz.eval_option('2', respuesta)

----------------------------------------
Tu respuesta: c, es correcta.
----------------------------------------


¿De qué tipo es la varible `nombre`?
* <class 'str'>
* <class 'complex'>
* <class 'float'>
* <class 'list'>
* <class 'int'>

In [7]:
# respuesta = "..."
### BEGIN SOLUTION
respuesta = "a"
file_answer.write('3', respuesta, 'Utiliza la función type(nombre) para revisar el tipo de la variable.')
### END SOLUTION

In [8]:
quizz.eval_option('3', respuesta)

----------------------------------------
Tu respuesta: a, es correcta.
----------------------------------------


# El huevo cocido perfecto

¿Cuál es el tiempo ideal para cocinar un huevo?

<img src="./huevo_cocido.jpg"  style="width: 300px;"/>


Para lograr un **huevo cocido** suave, la clara debe haberse calentado el tiempo suficiente para coagular a una temperatura superior a $63^oC$, pero la yema no debe calentarse por encima de $70^oC$. Para lograr un **huevo duro**, el centro de la yema debe de alcanzar los $70^oC$.

## Fórmula para calcular el tiempo de cocción.
La siguiente fórmula expresa el tiempo $t$, en segundos, que le toma a la yema alcanzar la temperatura $T_y$, en grados Celsius.

$$t = \dfrac{M^{2/3} c \rho^{1/3}}{K \pi^2 (4\pi/3)^{2/3}} \ln \left[ 0.76 \dfrac{T_o - T_w}{T_y - T_w}\right] \tag{1}$$

donde las propiedades son:

- $M$ masa; 
- $\rho$ densidad; 
- $c$ capacidad calorífica específica; 
- $K$ conductividad térmica; 
- $T_w$ es la temperatura de ebullición del agua;
- $T_o$ es la temperatura original del huevo antes de meterlo al agua;
- $T_y$ es la temperatura que debe alcanzar la yema.

**A Primer on Scientific Programming with Python**, Langtangen, H.P., ISBN 9783642183669, Texts in Computational Science and Engineering, https://books.google.com.co/books?id=Hi1KVomG148C, 2011, Springer Berlin Heidelberg.

<div class="alert alert-success">

## **Ejercicio 2.**

<font color="DarkBlue">
Calcular el tiempo de cocción necesario para un huevo duro pequeño de $M$ = 47 g para cuando la temperatura inicial 
del huevo está a temperatura ambiente: $T_{o} = 20^oC$.

**Datos**:
- $M$ = 47 g 
- $\rho$ = 1.038 g / cm$^3$
- $c$ = 3.7 J / g K
- $K$ = 5.4 $\times 10^{-3}$ W / cm K
- $T_w = 100^oC$
- $T_y = 70^oC$
  
</font>
</div>


### Paso 1.
Implementa $A = M^{2/3} c \rho^{1/3}$

**Nota**: para elevar un número a una potencia, $a^n$, podemos usar el operador `**` como sigue: `a**n` . Los números $a$ y $n$ pueden ser enteros o reales.

In [9]:
# Define las variables y realiza el cálculo
# M = ..
# ...
# A = math.pow(M,2/3) * ...

### BEGIN SOLUTION
# Definimos las variables
M   = 47    # Masa
rho = 1.038 # Densidad
c   = 3.7   # Capacidad calorífica específica

A = M**(2/3) * c * rho**(1/3)
file_answer.write('4', A, 'Revisa la precedencia de operadores, en caso de duda utiliza () cuando sea necesario, particularmente al calcular potencias.')
### END SOLUTION

print(A)

48.790216719661984


In [10]:
quizz.eval_numeric('4', A)

----------------------------------------
Tu resultado es correcto.
----------------------------------------


### Paso 2.
Implementa $B = K \pi^2 (4\pi/3)^{2/3}$

**Nota**: puedes obtener el valor de $\pi$ de la biblioteca `math` como sigue:
`from math import pi`.

In [11]:
### BEGIN SOLUTION
from math import pi

# Definimos las variables
K = 5.4e-3  # Conductividad térmica

B = K * pi**2 * (4 * pi / 3)**(2/3)
file_answer.write('5', B, 'Revisa la precedencia de operadores, en caso de duda utiliza () cuando sea necesario, particularmente al calcular potencias.')
### END SOLUTION
print(B)

0.13849026450902358


In [12]:
quizz.eval_numeric('5', B)

----------------------------------------
Tu resultado es correcto.
----------------------------------------


### Paso 3.
Implementa $C = \ln \left[ 0.76 \dfrac{T_o - T_w}{T_y - T_w}\right]$

**Nota**: en este caso se requiere de la función logaritmo natural, la cual puede ser obtenida del módulo `math` como sigue: `from math import log as ln`

In [13]:
### BEGIN SOLUTION
from math import log as ln

# Definimos las variables
To = 20  # Temperatura inicial del huevo
Tw = 100 # Temperatura de ebullición del agua
Ty = 70  # Temperatura que debe alcanzar la yema

C = ln(0.76 * (To - Tw) / (Ty - Tw))
file_answer.write('6', C, 'Revisa la precedencia de operadores, en caso de duda utiliza () cuando sea necesario.')
### END SOLUTION
print(C)

0.7063924073099658


In [14]:
quizz.eval_numeric('6', C)

----------------------------------------
Tu resultado es correcto.
----------------------------------------


### Paso 4.
Implementa el cálculo final: $t = \dfrac{A}{B} * C$. 

In [15]:
### BEGIN SOLUTION
t = A / B * C
file_answer.write('7', t, 'Revisa que todos los cálculos anteriores se hayan realizado correctamente.')
### END SOLUTION
print(t)

248.86253747844736


In [16]:
quizz.eval_numeric('7', t)

----------------------------------------
Tu resultado es correcto.
----------------------------------------


<div class="alert alert-success">

## **Ejercicio 3.**

<font color="DarkBlue">
Implementa la fórmula $(1)$ en una sola línea y realiza el cálculo para el mismo huevo pero para una temperatura: $T_{o} = 4^oC$ (acabado de sacar del refri).

</font>

</div>


In [17]:
### BEGIN SOLUTION
# Definimos las variables
To = 4  # Temperatura inicial del huevo
t = M**(2/3) * c * rho**(1/3) / ( K * pi**2 * (4 * pi / 3)**(2/3) ) * ln(0.76 * (To - Tw) / (Ty - Tw))
file_answer.write('8', t, 'Revisa la precedencia de operadores, en caso de duda utiliza () cuando sea necesario, particularmente al calcular potencias.')
### END SOLUTION

print(t)

313.09454902221637


In [18]:
quizz.eval_numeric('8', t)

----------------------------------------
Tu resultado es correcto.
----------------------------------------


<div class="alert alert-success">

## **Ejercicio 4.**

<font color="DarkBlue">
Implementa una función de Python que reciba como argumentos todos los valores que necesita la fórmula $(1)$ y que regrese como resultado el tiempo de cocción del huevo. 
    
Firma de la función:

`def calc_tiempo(M, To, c, rho, K, Tw, Ty):`

Ejecuta la función como sigue:

```python
To = 20
t1 = calc_tiempo(M, To, c, rho, K, Tw, Ty)

To = 4
t2 = calc_tiempo(M, To, c, rho, K, Tw, Ty)

print(t1, t2)
```

</font>
</div>

In [19]:
### BEGIN SOLUTION
def calc_tiempo(M, To, c, rho, K, Tw, Ty):
    return M**(2/3) * c * rho**(1/3) / ( K * pi**2 * (4 * pi / 3)**(2/3) ) * ln(0.76 * (To - Tw) / (Ty - Tw))

To = 20
t1 = calc_tiempo(M, To, c, rho, K, Tw, Ty)

To = 4
t2 = calc_tiempo(M, To, c, rho, K, Tw, Ty)
file_answer.write('9', [t1, t2], 'Copia correctamente el cálculo de la fórmula, del ejercicio anterior. Revisa como se debe regresar un valor calculado en un función.')
### END SOLUTION

print(t1, t2)


248.86253747844736 313.09454902221637


In [20]:
quizz.eval_numeric('9', [t1, t2])

----------------------------------------
Tu resultado es correcto.
----------------------------------------


<div class="alert alert-success">

## **Ejercicio 5.**

<font color="DarkBlue">

Realiza lo siguiente:
* Crea la función `calc_tiempo_np()` copiando el código completo de la función `calc_tiempo()`.
* En la nueva función `calc_tiempo_np()` modifica la parte del $\ln()$ para que se calcule usando la función `np.log()` de la biblioteca `numpy`.
* Posteriormente, usando `numpy`, crea el arreglo `Tol` con 9 valores de temperaturas en el intervalo $[4^o C, 20^o C]$, el arreglo debe contener los valores: `[ 4.,  6.,  8., 10., 12., 14., 16., 18., 20.]`. 
* Ejecuta la función como sigue: `tl = calc_tiempo_np(...)` pasando el arreglo `Tol`, el resultado debe ser un arreglo de 9 tiempos de cocción para cada una de las temperaturas del arreglo `Tol`, y debe contener algo similar a lo siguiente: `[313.09454902 305.67741829 ... 248.86253748]`

**Nota**: en este caso, cálculo de $\dfrac{T_o - T_w}{T_y - T_w}$ se hará con arreglos de `numpy`, por esta razón, tanto $T_w$ como $T_y$ también deben ser arreglos al momento de ejecutar la función `calc_tiempo_np()`.

**Hint**: define también los arreglos `Twl` y `Tyl` del mimso tamaño que `Tol` y úsalos cuando ejecutes la función `calc_tiempo_np()`.

</font>
</div>

In [21]:
### BEGIN SOLUTION
import numpy as np

def calc_tiempo(M, To, c, rho, K, Tw, Ty):
    return M**(2/3) * c * rho**(1/3) / ( K * pi**2 * (4 * pi / 3)**(2/3) ) * np.log(0.76 * (To - Tw) / (Ty - Tw))

Tol = np.linspace(4,20,9)
Twl = np.ones(9) * Tw
Tyl = np.ones(9) * Ty

tl = calc_tiempo(M, Tol, c, rho, K, Twl, Tyl)
file_answer.write('10', tl, 'Revisa con cuidado la descripción del ejercicio, así como la Nota y el Hint.')
file_answer.to_file('q1')
### END SOLUTION
print(Tol)
print(tl)

El directorio :/home/jovyan/HeCompA/.ans/01_Introduccion/ ya existe
Respuestas y retroalimentación almacenadas.
[ 4.  6.  8. 10. 12. 14. 16. 18. 20.]
[313.09454902 305.67741829 298.1007672  290.35758304 282.44038026
 274.34115696 266.05134632 257.56176211 248.86253748]


In [22]:
quizz.eval_numeric('10', tl)

----------------------------------------
Tu resultado es correcto.
----------------------------------------
